include/criterion/internal/assert.h

```diff
@@ -128,16 +144,33 @@ CR_END_C_API
             Fail();                                                         \
     } while (0)
 
-# define cr_fail(Fail, ...)                                     \
-    CR_EXPAND(cr_assert_impl(                                   \
-            Fail,                                               \
-            0,                                                  \
-            dummy,                                              \
-            CRITERION_ASSERT_MSG_FAIL,                          \
-            (),                                                 \
-            __VA_ARGS__                                         \
+
+# define cr_skip_impl(Skip, ...)                                            \
+    do {                                                                    \
+        char *cr_msg__ = NULL;                                              \
+        int cr_shifted__ = 0;                                               \
+        CR_EXPAND(CR_INIT_SKIP_MSG(cr_msg__, cr_shifted__, __VA_ARGS__));   \
+        Skip(cr_msg__);                                                     \
+        CR_STDN free(cr_msg__ - cr_shifted__);                              \
```

Comment:
This leaks as `Skip` never returns. This is not solvable as-is, you have to get rid of `cr_skip_impl` entirely as it's overkill for what it does (the only reason this code works for the assert macros is that they also take a default format string and parameters).

I would still implement the optional string as a `"" __VA_ARGS__` as above but make `criterion_skip_test` variadic. The prototype of the function would be:

`CR_API CR_NORETURN void criterion_skip_test(const char *format, ...);`.

Another possibility would be to remove `criterion_skip_test` from the API altogether and make it internal. This way, you have the liberty of freeing the message from the function itself -- but I still think that cr_skip_impl is a bit too complicated for what it really is and that making the function variadic would work best for this purpose.